File: internal/telemetry/middleware.go
package telemetry

import (
	"context"
	"net/http"
	"regexp"
	"strconv"
	"time"

	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"
)

// HTTPMiddleware provides OpenTelemetry instrumentation for HTTP handlers
type HTTPMiddleware struct {
	provider         *Provider
	excludedPatterns []*regexp.Regexp
	serviceName      string
}

// NewHTTPMiddleware creates a new HTTP middleware with telemetry
func NewHTTPMiddleware(provider *Provider, serviceName string) *HTTPMiddleware {
	return &HTTPMiddleware{
		provider:    provider,
		serviceName: serviceName,
		excludedPatterns: []*regexp.Regexp{
			regexp.MustCompile(`/-/health/?$`),
			regexp.MustCompile(`/-/metrics/?$`),
			regexp.MustCompile(`/-/ready/?$`),
			regexp.MustCompile(`/-/live/?$`),
			regexp.MustCompile(`/static/.*`),
			regexp.MustCompile(`/favicon\.ico$`),
		},
	}
}

// Handler wraps an HTTP handler with OpenTelemetry instrumentation
func (m *HTTPMiddleware) Handler(next http.Handler) http.Handler {
	if !m.provider.IsEnabled() {
		return next
	}

	// Use otelhttp for automatic instrumentation
	return otelhttp.NewHandler(
		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip excluded paths
			if m.shouldExclude(r.URL.Path) {
				next.ServeHTTP(w, r)
				return
			}

			// Add custom span attributes
			span := trace.SpanFromContext(r.Context())
			m.setSpanAttributes(span, r)

			// Record request start time
			start := time.Now()

			// Create response writer wrapper to capture status code
			rw := &responseWriter{ResponseWriter: w, statusCode: 200}

			// Process request
			next.ServeHTTP(rw, r)

			// Record metrics
			duration := time.Since(start).Seconds()
			status := strconv.Itoa(rw.statusCode)

			m.provider.RecordRequest(r.Context(), r.Method, status, duration)

			// Set span status for errors
			if rw.statusCode >= 400 {
				span.SetAttributes(attribute.Int("http.status_code", rw.statusCode))
				if rw.statusCode >= 500 {
					span.RecordError(nil, trace.WithAttributes(
						attribute.String("error.type", "http_server_error"),
						attribute.String("error.message", http.StatusText(rw.statusCode)),
					))
				}
			}
		}),
		"http.request",
		otelhttp.WithSpanNameFormatter(m.getSpanName),
	)
}

// HandlerFunc wraps an HTTP handler function with OpenTelemetry instrumentation
func (m *HTTPMiddleware) HandlerFunc(next http.HandlerFunc) http.Handler {
	return m.Handler(next)
}

// shouldExclude checks if a request path should be excluded from tracing
func (m *HTTPMiddleware) shouldExclude(path string) bool {
	for _, pattern := range m.excludedPatterns {
		if pattern.MatchString(path) {
			return true
		}
	}
	return false
}

// setSpanAttributes sets custom attributes on the span
func (m *HTTPMiddleware) setSpanAttributes(span trace.Span, r *http.Request) {
	if span == nil || !span.IsRecording() {
		return
	}

	// Set service-specific attributes
	span.SetAttributes(
		attribute.String("service.name", m.serviceName),
		attribute.String("http.method", r.Method),
		attribute.String("http.url", r.URL.String()),
		attribute.String("http.scheme", r.URL.Scheme),
		attribute.String("http.host", r.Host),
		attribute.String("http.target", r.URL.Path),
	)

	// User agent
	if userAgent := r.Header.Get("User-Agent"); userAgent != "" {
		span.SetAttributes(attribute.String("http.user_agent", userAgent))
	}

	// Client IP
	if clientIP := getClientIP(r); clientIP != "" {
		span.SetAttributes(attribute.String("http.client_ip", clientIP))
	}

	// Content length
	if r.ContentLength > 0 {
		span.SetAttributes(attribute.Int64("http.request_content_length", r.ContentLength))
	}
}

// getSpanName generates a span name for the request
func (m *HTTPMiddleware) getSpanName(operation string, r *http.Request) string {
	method := r.Method
	path := r.URL.Path

	// Simplify common API patterns
	if path == "/" {
		return method + " /"
	}

	// For outpost endpoints, use generic names
	switch {
	case regexp.MustCompile(`^/outpost\.goauthentik\.io/`).MatchString(path):
		return method + " /outpost.goauthentik.io/*"
	case regexp.MustCompile(`^/akprox/`).MatchString(path):
		return method + " /akprox/*"
	case regexp.MustCompile(`^/api/`).MatchString(path):
		return method + " /api/*"
	default:
		return method + " " + path
	}
}

// getClientIP extracts the client IP from the request
func getClientIP(r *http.Request) string {
	// Check X-Forwarded-For header
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		return xff
	}

	// Check X-Real-IP header
	if xri := r.Header.Get("X-Real-IP"); xri != "" {
		return xri
	}

	// Fall back to RemoteAddr
	return r.RemoteAddr
}

// responseWriter wraps http.ResponseWriter to capture the status code
type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

// WriteHeader captures the status code
func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// Write ensures WriteHeader is called
func (rw *responseWriter) Write(b []byte) (int, error) {
	if rw.statusCode == 0 {
		rw.statusCode = 200
	}
	return rw.ResponseWriter.Write(b)
}

// LDAPMiddleware provides telemetry for LDAP operations
type LDAPMiddleware struct {
	provider *Provider
}

// NewLDAPMiddleware creates a new LDAP middleware with telemetry
func NewLDAPMiddleware(provider *Provider) *LDAPMiddleware {
	return &LDAPMiddleware{
		provider: provider,
	}
}

// RecordOperation records an LDAP operation with telemetry
func (m *LDAPMiddleware) RecordOperation(ctx context.Context, operation string, fn func() error) error {
	if !m.provider.IsEnabled() {
		return fn()
	}

	// Start tracing span
	tracer := m.provider.GetTracer("authentik.ldap")
	ctx, span := tracer.Start(ctx, "ldap."+operation)
	defer span.End()

	// Set span attributes
	span.SetAttributes(
		attribute.String("ldap.operation", operation),
	)

	// Record operation start time
	start := time.Now()

	// Execute operation
	err := fn()

	// Record metrics
	duration := time.Since(start).Seconds()
	success := err == nil

	m.provider.RecordLDAPOperation(ctx, operation, duration, success)

	// Record error if any
	if err != nil {
		span.RecordError(err)
		span.SetAttributes(attribute.Bool("error", true))
	}

	return err
}

// AuthMiddleware provides telemetry for authentication operations
type AuthMiddleware struct {
	provider *Provider
}

// NewAuthMiddleware creates a new authentication middleware with telemetry
func NewAuthMiddleware(provider *Provider) *AuthMiddleware {
	return &AuthMiddleware{
		provider: provider,
	}
}

// RecordAuth records an authentication attempt with telemetry
func (m *AuthMiddleware) RecordAuth(ctx context.Context, method string, fn func() (bool, error)) (bool, error) {
	if !m.provider.IsEnabled() {
		return fn()
	}

	// Start tracing span
	tracer := m.provider.GetTracer("authentik.auth")
	ctx, span := tracer.Start(ctx, "auth."+method)
	defer span.End()

	// Set span attributes
	span.SetAttributes(
		attribute.String("auth.method", method),
	)

	// Execute authentication
	success, err := fn()

	// Record metrics
	m.provider.RecordAuth(ctx, method, success)

	// Set span attributes based on result
	span.SetAttributes(
		attribute.Bool("auth.success", success),
	)

	// Record error if any
	if err != nil {
		span.RecordError(err)
		span.SetAttributes(attribute.Bool("error", true))
	}

	return success, err
}

File: authentik/lib/telemetry/middleware.py
"""OpenTelemetry middleware for Django"""

import re
import time
import logging
from typing import Callable, Optional

from django.http import HttpRequest, HttpResponse
from django.utils.deprecation import MiddlewareMixin

from authentik.lib.telemetry import provider

logger = logging.getLogger(__name__)


class OpenTelemetryMiddleware(MiddlewareMixin):
    """Django middleware for OpenTelemetry tracing"""

    def __init__(self, get_response: Callable):
        super().__init__(get_response)
        self.get_response = get_response
        self.tracer = None
        self.sampler = provider.AdaptiveSampler()

        # Patterns for requests to exclude from tracing
        self.excluded_patterns = [
            re.compile(r"/health/?$"),
            re.compile(r"/metrics/?$"),
            re.compile(r"/readiness/?$"),
            re.compile(r"/liveness/?$"),
            re.compile(r"/static/.*"),
            re.compile(r"/media/.*"),
        ]

        # Initialize tracer if telemetry is enabled
        if provider.provider.is_enabled():
            provider.provider.initialize()
            self.tracer = provider.provider.get_tracer("authentik.middleware")

    def __call__(self, request: HttpRequest) -> HttpResponse:
        """Process request with OpenTelemetry tracing"""
        if not self.tracer or self._should_exclude_request(request):
            return self.get_response(request)

        # Create span for the request
        span_name = self._get_span_name(request)

        with self.tracer.start_as_current_span(span_name) as span:
            # Record request attributes
            self._set_request_attributes(span, request)

            start_time = time.time()
            try:
                response = self.get_response(request)

                # Record response attributes
                self._set_response_attributes(span, response)

                return response

            except Exception as exc:
                # Record exception details
                span.record_exception(exc)
                span.set_status(self._get_error_status(exc))
                raise

            finally:
                # Record request duration
                duration = time.time() - start_time
                span.set_attribute("http.response_time", duration)

    def _should_exclude_request(self, request: HttpRequest) -> bool:
        """Check if request should be excluded from tracing"""
        path = request.path_info

        for pattern in self.excluded_patterns:
            if pattern.match(path):
                return True

        return False

    def _get_span_name(self, request: HttpRequest) -> str:
        """Generate span name for the request"""
        method = request.method
        path = request.path_info

        # Simplify common API patterns
        if path.startswith("/api/v3/"):
            # Extract API endpoint
            parts = path.split("/")
            if len(parts) >= 4:
                return f"{method} /api/v3/{parts[3]}"

        return f"{method} {path}"

    def _set_request_attributes(self, span, request: HttpRequest):
        """Set request-related span attributes"""
        try:
            span.set_attribute("http.method", request.method)
            span.set_attribute("http.url", request.build_absolute_uri())
            span.set_attribute("http.route", request.path_info)
            span.set_attribute("http.scheme", request.scheme)

            # User agent
            user_agent = request.META.get("HTTP_USER_AGENT")
            if user_agent:
                span.set_attribute("http.user_agent", user_agent)

            # Client IP
            client_ip = self._get_client_ip(request)
            if client_ip:
                span.set_attribute("http.client_ip", client_ip)

            # User information if available
            if hasattr(request, "user") and request.user and hasattr(request.user, "pk"):
                if request.user.is_authenticated:
                    span.set_attribute("user.id", str(request.user.pk))
                    span.set_attribute("user.username", getattr(request.user, "username", ""))

        except Exception as exc:
            logger.warning("Failed to set request attributes", exc_info=exc)

    def _set_response_attributes(self, span, response: HttpResponse):
        """Set response-related span attributes"""
        try:
            span.set_attribute("http.status_code", response.status_code)

            # Set status based on HTTP status code
            if response.status_code >= 400:
                span.set_status(self._get_error_status_from_code(response.status_code))

            # Content type
            content_type = response.get("Content-Type", "")
            if content_type:
                span.set_attribute("http.response.content_type", content_type)

        except Exception as exc:
            logger.warning("Failed to set response attributes", exc_info=exc)

    def _get_client_ip(self, request: HttpRequest) -> Optional[str]:
        """Extract client IP from request"""
        # Check X-Forwarded-For header
        x_forwarded_for = request.META.get("HTTP_X_FORWARDED_FOR")
        if x_forwarded_for:
            return x_forwarded_for.split(",")[0].strip()

        # Check X-Real-IP header
        x_real_ip = request.META.get("HTTP_X_REAL_IP")
        if x_real_ip:
            return x_real_ip

        # Fall back to REMOTE_ADDR
        return request.META.get("REMOTE_ADDR")

    def _get_error_status(self, exc: Exception):
        """Get OpenTelemetry status for an exception"""
        try:
            from opentelemetry.trace import Status, StatusCode

            return Status(StatusCode.ERROR, str(exc))
        except ImportError:
            return None

    def _get_error_status_from_code(self, status_code: int):
        """Get OpenTelemetry status for HTTP status code"""
        try:
            from opentelemetry.trace import Status, StatusCode

            if status_code >= 500:
                return Status(StatusCode.ERROR, f"HTTP {status_code}")
            elif status_code >= 400:
                return Status(StatusCode.ERROR, f"HTTP {status_code}")
            return Status(StatusCode.OK)
        except ImportError:
            return None
